Libraries/oneMKL/fourier_correlation/fcorr_1d_usm.cpp

```diff
@@ -13,103 +13,189 @@
+    sig1[N - N / 4 + 1] = 1.0f;
+    sig2[N / 4 - 1]     = 1.0f;
+    sig2[N / 4]         = 1.0f;
+    sig2[N / 4 + 1]     = 1.0f;
+  });
+  // Calculate L2 norms of both input signals before proceeding (for
+  // normalization purposes and for the definition of error tolerance)
+  float *norm_sig1 = sycl::malloc_shared<float>(1, Q);
+  float *norm_sig2 = sycl::malloc_shared<float>(1, Q);
+  evt1 = oneapi::mkl::blas::nrm2(Q, N, sig1, 1, norm_sig1, {evt});
+  evt2 = oneapi::mkl::blas::nrm2(Q, N, sig2, 1, norm_sig2, {evt});
+  // 1) Calculate the cross-correlation naively (for verification purposes);
+  naive_cross_correlation(Q, N, sig1, sig2, naive_corr, {evt}).wait();
+  // 2) Calculate the cross-correlation via Discrete Fourier Transforms (DFTs):
+  //       corr = (1/N) * iDFT(DFT(sig1) * CONJ(DFT(sig2)))
+  // Initialize DFT descriptor
   oneapi::mkl::dft::descriptor<oneapi::mkl::dft::precision::SINGLE,
-                               oneapi::mkl::dft::domain::REAL>
-      transform_plan(N);
-  transform_plan.commit(Q);
-
-  // Perform forward transforms on real arrays
-  evt1 = oneapi::mkl::dft::compute_forward(transform_plan, sig1);
-  evt2 = oneapi::mkl::dft::compute_forward(transform_plan, sig2);
-
-  // Compute: DFT(sig1) * CONJG(DFT(sig2))
-  oneapi::mkl::vm::mulbyconj(
-      Q, N / 2, reinterpret_cast<std::complex<float>*>(sig1),
-      reinterpret_cast<std::complex<float>*>(sig2),
-      reinterpret_cast<std::complex<float>*>(corr), {evt1, evt2})
-      .wait();
-
-  // Perform backward transform on complex correlation array
-  oneapi::mkl::dft::compute_backward(transform_plan, corr).wait();
-
-  // Find the shift that gives maximum correlation value using parallel maxloc
-  // reduction
-  std::pair<float, int>* max_res =
-       sycl::malloc_shared<std::pair<float, int>>(1, Q);
-  std::pair<float, int> max_identity = {std::numeric_limits<float>::min(),
-                                        std::numeric_limits<int>::min()};
-  *max_res = max_identity;
-  auto red_max = reduction(max_res, max_identity, maxloc<float, int>());
-
-  Q.parallel_for(sycl::nd_range<1>{N, L}, red_max,
-                 [=](sycl::nd_item<1> item, auto& max_res) {
-                    int i = item.get_global_id(0);
-		    std::pair<float, int> partial = {corr[i], i};
-                    max_res.combine(partial);
-   })
-      .wait();
-  float max_corr = max_res->first;
-  int shift = max_res->second;
-  shift =
-      (shift > N / 2) ? shift - N : shift;  // Treat the signals as circularly
-                                            // shifted versions of each other.
-  std::cout << "Shift the second signal " << shift
-            << " elements relative to the first signal to get a maximum, "
-               "normalized correlation score of "
-            << max_corr / N << ".\n";
+                               oneapi::mkl::dft::domain::REAL> desc(N);
+  desc.set_value(oneapi::mkl::dft::config_param::BACKWARD_SCALE, 1.0f / N);
```

Comment:
Same comment as before for forward scale

Reply:
Done in e776e50b2647a4e7d10808cd8a9c2617005766db